### ISSUE TITLE
Use unsigned long for axis of concat operation

### ISSUE THREAD

**Issue by miaobin** [webmachinelearning/webnn#345]
```
partial interface MLGraphBuilder {
  MLOperand concat(sequence<MLOperand> inputs, long axis);
};
```
We now limit the value of [axis](https://webmachinelearning.github.io/webnn/#dom-mlgraphbuilder-concat-inputs-axis-axis) in the interval [0, N) where N is the rank of all the inputs. It should be a positive value and we may use `unsigned long` for axis.

**Comment by huningxin** [comment]
@miaobin , thanks for raising this issue. Given the valid value of `axis` is in the range [0, N), it makes sense to me that it should be of type `unsigned long`. 

And there may be an opportunity to unify the `axis` definitions across this spec. For example, `MLSplitOptions.axis` is also `long`, and the spec says "A negative value is interpreted as counting back from the end.". This is inconsistent with `concat`.

From another point of view, because WebNN [dropped](https://github.com/webmachinelearning/webnn/pull/281) the support of dynamic tensor, it would make the negative axis less useful.

The `unsigned long` axis would also simplify the validation code and mapping to native ML API. e.g., DirectML's [`DML_JOIN_OPERATOR_DESC`](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_join_operator_desc) and [`DML_SPLIT_OPERATOR_DESC`](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_split_operator_desc) declare `axis` as `UINT`, XNNPACK's [`xnn_define_concatenate2()`](https://github.com/google/XNNPACK/blob/master/include/xnnpack.h#LL949C17-L949C40) and [`xnn_define_even_split2`()`](https://github.com/google/XNNPACK/blob/master/include/xnnpack.h#LL1047C17-L1047C39) declare `axis` as `size_t`.

So, I propose to define the `axis` to be type of `unsigned long` and in the range of [0, N) where N is the rank of the tensor.

**Comment by BruceDai** [comment]
> And there may be an opportunity to unify the axis definitions across this spec.

Link to #307, [batchNormalization](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-batchnorm) also uses `axis` option.

**Comment by huningxin** [comment]
Related to Chromium `concat` [CL](https://chromium-review.googlesource.com/c/chromium/src/+/4174743).

**Comment by wacky6** [comment]
There're a few things to consider in WebNN context:

* Intended user: do we expect developers use this API directly (where negative indices makes more sense), or via a framework (e.g. tfjs, which can easily figure out the correct index before calling this API)
* Implementation complexity and robustness: signed integer needs extra caution during validation stage. Existing CL's CheckedNumeric + `if x<0` takes more mental energy to understand than unsigned math (Do the math, then AssignIfValid)
* The last dimension usage. I think this is related to static / dynamic graph issue? If we only support static graph (i.e. can't change tensor shape after graph initialization), then all dimensions are known at graph build time. In this case negative index is a mere syntax sugar at the cost of more complicated / error-prone C++ math calculation, and more WPT test cases

**Comment by huningxin** [comment]
@wacky6 , thanks for your implementation reviewer perspectives, very helpful!

> * Intended user: 

WebNN aims at being framework backend. So the users of this API would be most likely the framework developers.  

I agree with you that framework can support negative axis in their way and does the conversion if the backend API only supports unsigned integer axis.

For example, although TFLite supports negative axis, XNNPACK only support unsigned integer axis. The TFLite XNNPACK delegate does the [conversion](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc#L2816) by `if (axis < 0) axis += rank`. As another example, ONNXRuntime DirectML execution provider uses [`GetDmlAdjustedAxis()`](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorConcat.cpp#L37) to convert negative axis supported by ONNX to unsigned integer axis that is supported by DirectML. 

So, I guess if WebNN supports unsigned integer axis, the WebNN framework backend cann implement in a similar way.

> * Implementation complexity and robustness:

I agree the unsigned mat would be much easier and less error-prone. 

If WebNN supports unsigned integer axis, when implement a backend that supports signed integer axis, such as MPSGraph, it may be cautious about overflow when converting unsigned integer to signed integer.

> * The last dimension usage

I agree it is less useful for static graph that WebNN only supports. 

My another point this spec should make the axis definition consistent across APIs, including @BruceDai mentioned in #307 for `batchNormalization`, and also @lisa0314 mentioned in #317 for `transpose`. I copied Bruce's table and add all ops that uses axis/axes.

**Comment by huningxin** [comment]
My another point this spec should make the axis definition consistent across APIs, including @BruceDai mentioned in #307 for `batchNormalization`, and also @lisa0314 mentioned in #317 for `transpose`. I copied Bruce's table and add all ops that uses axis/axes.

|Op | Axis | Type| Value description|
|----|----|----|----|
|[batchNormalization](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-batchnorm)| `MLBatchNormalizationOptions.axis`  |long|When it’s not specified, the default value is 1.When input is a 4-D tensor of the "nchw" or "nhwc" layout, options.axis should be set to 1 or 3 respectively. |
|[concat](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-concat)|`axis` is the second parameter| long|with the value in the interval [0, N) where N is the rank of all the inputs|
|[split](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-split)|`MLSplitOptions.axis`|long|Default to 0. A negative value is interpreted as counting back from the end|
|[Reduction operations](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-reduce) | `MLReduceOptions.axes` | `sequence<long>` | axes: a sequence of long. The dimensions to reduce where -1 means the last dimension. If not present, all dimensions are reduced. |
| [resample2d](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-resample2d) | `MLResample2dOptions.axes` | `sequence<long>` | axes: a sequence of long of length 2. The two consecutive dimensions of the input tensor to which the interpolation algorithm applies. The valid values in the sequence are [0, 1], [1, 2] or [2, 3]. When not specified, the sequence is assumed to be [2, 3]. |
| [slice](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-slice) | `MLSliceOptions.axes` | `sequence<long>` | axes: a sequence of long. The dimensions of the input shape to which starts and sizes apply. The values in the sequence are either within the [0, r-1] range where r is the input tensor rank, or the [-r, -1] range where negative values mean counting back from the end of the input shape. When not specified, the sequence is assumed to be [0,1,..r-1]. |
| [squeeze](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-squeeze) | `MLSqueezeOptions.axes`  |  `sequence<long>` | axes: a sequence of [long](https://webidl.spec.whatwg.org/#idl-long). Indices to the shape dimensions of size 1 to eliminate. When not specified, every shape dimensions of size 1 in the tensor are eliminated. |
| [transpose](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-transpose) | `MLTransposeOptions.permutation` |  `sequence<long>` | permutation: a sequence of long values. The values used to permute the output shape. When it’s not specified, it’s set to [N-1...0], where N is the rank of the input tensor. These default values cause the output to become a transposed tensor of the input. When specified, the number of values in the sequence must be the same as the rank of the input tensor, and the values in the sequence must be within the range from 0 to N-1 with no two or more same values found in the sequence. |

**Comment by wchao1115** [comment]
@fdwr

**Comment by fdwr** [comment]
@huningxin Thanks for the holistic table above comparing operators. Yeah, from the *consistency* POV, I strongly agree that all the operators sharing similar indices/axes either all share this special policy of negative numbers or that they all do not (so if `split` supports negative axes, then `concat` should too, or vice versa). I don't see the negative number policy is that convenient anyway given other existing constraints in WebNN like ahead-of-time shape inference, and I'm not sure the original target audience for WebNN (more data scientists vs more framework implementers), but my view of WebNN has been that it is lower level than say PyTorch or TensorFlow, and should be more explicit, leaving higher level framework policies like that to already be resolved before reaching WebNN. This also makes testing WebNN conformance simpler.

The one case where I would argue in *favor* of negative axes for generality is if your model supported arbitrary rank inputs (e.g. although uncommon, ONNX models allow arbitrary rank inputs 1D/2D/3D... by just [leaving input TensorProto shape blank](https://github.com/onnx/onnx/blob/main/docs/ShapeInference.md#background), and then from-the-end indexing is useful), but again since WebNN requires statically known shapes ahead of time, there should be no ambiguous cases where negative numbers can't be resolved ahead of time.

**Comment by huningxin** [comment]
@fdwr , great points!

> This also makes testing WebNN conformance simpler.

+1. And I think WebNN test developers, @BruceDai , would be happy about that.

With inputs from both @fdwr and @wacky6 , I'd drop PR #352 and make another PR that unifies the axis/axes definitions by using unsinged integer with a consistent value range. I'll invite you review once it is ready.

Thanks!

**Comment by huningxin** [comment]
The PR #359 that uses unsigned integer axis is ready for review. Thanks!
